File: algorithms/js/63.不同路径-ii.js
/*
 * @lc app=leetcode.cn id=63 lang=javascript
 *
 * [63] 不同路径 II
 */

/**
 * bottom up
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {
  const row = obstacleGrid.length;
  const col = obstacleGrid[0].length;
  if (obstacleGrid[row - 1][col - 1] || obstacleGrid[0][0]) return 0;
  const dp = Array.from({ length: col }, () => 1);

  for (let i = row - 1; i > -1; i--) {
    for (let j = col - 1; j > -1; j--) {
      if (i === row - 1)
        dp[j] = obstacleGrid[i][j] == 0 && dp[j + 1] !== 0 ? 1 : 0;
      else dp[j] = obstacleGrid[i][j] == 1 ? 0 : dp[j] + (dp[j + 1] || 0);
    }
  }
  return dp[0];
};

// @lc code=start
/**
 * up bottom
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {
  const row = obstacleGrid.length;
  const col = obstacleGrid[0].length;
  if (obstacleGrid[row - 1][col - 1] === 1) return 0;
  const dp = Array.from({ length: col }, () => 0);
  dp[0] = 1;

  for (let i = 0; i < row; i++) {
    for (let j = 0; j < col; j++) {
      if (obstacleGrid[i][j] === 1) dp[j] = 0;
      else dp[j] += dp[j - 1] || 0;
    }
  }
  return dp[col - 1];
};

File: algorithms/golang/63.不同路径-ii.go
/*
 * @lc app=leetcode.cn id=63 lang=golang
 *
 * [63] 不同路径 II
 */

// @lc code=start
/**
 * DP, 自底向上
 * DP方程: opt[i][j]=opt[i+1][j]+opt[i][j+1]
 */
func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	row, col := len(obstacleGrid), len(obstacleGrid[0])
	if obstacleGrid[row-1][col-1] == 1 {
		return 0
	}
	opt := make([][]int, row)
	for i := 0; i < row; i++ {
		opt[i] = make([]int, col)
	}
	opt[row-1][col-1] = 1
	for i := row - 1; i > -1; i-- {
		for j := col - 1; j > -1; j-- {
			if obstacleGrid[i][j] == 1 {
				opt[i][j] = 0
				continue
			}
			if j+1 < col && obstacleGrid[i][j+1] == 0 {
				opt[i][j] += opt[i][j+1]
			}
			if i+1 < row && obstacleGrid[i+1][j] == 0 {
				opt[i][j] += opt[i+1][j]
			}

		}
	}
	return opt[0][0]
}

// @lc code=end